programs/redstone-sol/src/util.rs

```diff
@@ -0,0 +1,121 @@
+pub type U256 = [u8; 32];
+
+pub trait Trim<T>
+where
+    Self: Sized,
+{
+    fn trim_end(&mut self, len: usize) -> T;
+}
+
+impl Trim<Vec<u8>> for Vec<u8> {
+    fn trim_end(&mut self, len: usize) -> Self {
+        if len >= self.len() {
+            std::mem::take(self)
+        } else {
+            self.split_off(self.len() - len)
+        }
+    }
+}
+
+pub fn u256_to_string(u256: U256) -> String {
+    u256.iter()
+        .take_while(|&&c| c != 0)
+        .map(|&c| c as char)
+        .collect()
+}
+
+pub fn u256_to_num_string(u256: U256) -> String {
+    let mut num = 0u128;
+    for &c in &u256 {
+        num = (num << 8) | c as u128;
+    }
+    num.to_string()
+}
+
+#[cfg(feature = "dev")]
+pub fn bytes_to_hex(bytes: &[u8]) -> String {
+    use std::fmt::Write;
+    bytes.iter().fold(
+        String::with_capacity(bytes.len() * 2),
+        |mut output, b| {
+            let _ = write!(output, "{:02x}", b);
+            output
+        },
+    )
+}
+
+pub trait FromBytesRepr<T: Sized>: Sized {
+    fn from_bytes(bytes: T) -> Self;
+}
+
+impl FromBytesRepr<&[u8]> for U256 {
+    fn from_bytes(bytes: &[u8]) -> Self {
+        let mut array = [0u8; 32];
+        let len = bytes.len().min(32);
+        array[..len].copy_from_slice(&bytes[..len]);
+        array
+    }
+}
+
+impl FromBytesRepr<&[u8]> for usize {
+    fn from_bytes(bytes: &[u8]) -> Self {
+        bytes.iter().fold(0usize, |acc, &b| (acc << 8) | b as usize)
+    }
+}
+
+impl FromBytesRepr<&[u8]> for u64 {
+    fn from_bytes(bytes: &[u8]) -> Self {
+        bytes.iter().fold(0u64, |acc, &b| (acc << 8) | b as u64)
+    }
+}
+
+impl FromBytesRepr<&[u8]> for u128 {
+    fn from_bytes(bytes: &[u8]) -> Self {
+        bytes.iter().fold(0u128, |acc, &b| (acc << 8) | b as u128)
+    }
+}
+
+pub fn calculate_median(values: &mut [U256]) -> U256 {
+    if values.is_empty() {
+        return [0u8; 32];
+    }
+
+    values.sort_unstable();
+    let len = values.len();
+
+    if len % 2 == 0 {
```

Comment:
still, [U256_MAX, U256_MAX, 2, 2] will return U256_MAX instead of (U256_MAX / 2 + 1) :(